### PR TITLE
feat(parser): allow to add identifiers

### PR DESCRIPTION
**Pull request by RomainLanz**
Hey there! 👋🏻

This PR is the continuation of https://github.com/adonisjs/env/pull/35.
It adds a way to set up an identifier for an environment variable, helping to interpolate the value of your env.

It can be useful in multiple scenarios.

1. To manage your secrets, you may use a Secrets Vault like `Docker Secret` or `HashiCorp Vault`. Doing so it will generate a path for your secret and won't give you any value directly. You can write a custom logic to read the file content.
```.env
DB_PASSWORD=file:/run/secrets/db_password
````

2. You may encrypt your secrets. You will be able to decrypt them on the fly.
```.env
APP_KEY=encrypted:ksadskjadsjkdasjkdsjdjaskjkdjkasdkj
```

```ts
Env.identifier('file', (value: string) => {
  // ...
})

Env.create(new URL('../', import.meta.url), {
  ....
})
```

### CODE REVIEW

File: src/parser.ts [hunk 1]
```diff
@@ -174,12 +195,37 @@ export class EnvParser {
   /**
    * Parse the env string to an object of environment variables.
    */
-  parse(): DotenvParseOutput {
+  async parse(): Promise<DotenvParseOutput> {
     const envCollection = dotenv.parse(this.#envContents.trim())
+    const identifiers = Object.keys(EnvParser.#identifiers)
+    let result: DotenvParseOutput = {}
+
+    $keyLoop: for (const key in envCollection) {
```
Comment (thetutlage) [inline src/parser.ts:203]: Not sure, I understand this `$keyLoop` syntax. Can't this code be as simple as

```ts
for (const key in envCollection) {
  const value = this.#getValue(key, envCollection)

  if (value.includes(':')) {
    for (const identifier of identifiers) {
      if (value.startsWith(`${identifier}\\:`)) {
        result[key] = identifier + value.substring(identifier.length + 1)
      } else if (value.startsWith(`${identifier}:`)) {
        result[key] = await EnvParser.#identifiers[identifier](
          value.substring(identifier.length + 1)
        )
      }
    }
    result[key] = value
  } else {
    result[key] = value
  }
}
```

Comment (RomainLanz) [reply src/parser.ts:203]: We have to break the outer loop.

In your code example, you are reassigning the `result[key]` array at the end of the second `for` loop.

```ts
for (const identifier of identifiers) {
  if (value.startsWith(`${identifier}\\:`)) {
    result[key] = identifier + value.substring(identifier.length + 1)
  } else if (value.startsWith(`${identifier}:`)) {
    result[key] = await EnvParser.#identifiers[identifier](
      value.substring(identifier.length + 1)
    )
  }
}
result[key] = value // 👈 Here the value is lost
```

A [JavaScript label](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/label) can be used to name a loop and use `continue` or `break` on it even if you are into another loop.

Comment (thetutlage) [reply src/parser.ts:203]: I mean, we can use the `break` statement without having the label.

To be honest, I have never used labels and neither they are anywhere in the codebase. Are they usually helpful when you are inside nested loops?

Comment (RomainLanz) [reply src/parser.ts:203]: > I mean, we can use the break statement without having the label.

We have to `continue` on outside loop, this is the issue.

> I have never used labels and neither they are anywhere in the codebase

If you prefer, we can make it work by storing a temporary variable that defines if the value has been fine and then continue the loop after. Something like:

```ts
for (const key in envCollection) {
  let valueFound = false // 👈
  const value = this.#getValue(key, envCollection)
  
  if (value.includes(':')) { 
    for (const identifier of identifiers) {
      if (value.startsWith(`${identifier}\\:`)) {
        result[key] = identifier + value.substring(identifier.length + 1)
        valueFound = true // 👈
        break
      }
      
      if (value.startsWith(`${identifier}:`)) {
        result[key] = await EnvParser.#identifiers[identifier](
          value.substring(identifier.length + 1)
        )
        valueFound = true // 👈
        break
      }
    }
    
    if (valueFound) { // 👈
      continue
    }
    
    result[key] = value
  } else {
    result[key] = value
  }
}
```

Comment (thetutlage) [reply src/parser.ts:203]: Okay, maybe I am missing something. Let's go with the label approach :)

**Comment by thetutlage** [comment]
A couple of things to keep in mind

- It is going to be a breaking change for this package.
- We will absorb the breaking change in the framework core and hence it can be minor release for `@adonisjs/core`.
- We should update the README file with the identifier usage
- And also update the docs with the identifier usage.

**Comment by RomainLanz** [comment]
> It is going to be a breaking change for this package.

Why is this a breaking change? The behavior is not changing for current user.

**Comment by thetutlage** [comment]
We have changed the parse method to async, which was sync earlier.

So anyone relying on this package using the Parser directly (which is publicly exported) has to now await the parse method call

**Comment by RomainLanz** [comment]
Ah yes, for anyone using the package directly, it is a breaking change, that's true. 😄

**Pull request merged**